File: packages/python/catalogmx/catalogs/sat/cfdi_4/estado.py
"""Catálogo c_Estado compartido por CFDI y Nómina 1.2."""

import json
from importlib.resources import files


class EstadoCfdiCatalog:
    """Estados y provincias publicados por SAT como c_Estado."""

    _data: list[dict] | None = None
    _by_code: dict[str, dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        if cls._data is None:
            # SAT defines c_Estado in catCFDI.xsd as an identifier set.  The
            # technical-mirror SQLite release is missing the DIF identifier,
            # so this catalog intentionally consumes the reviewed XSD-derived
            # package artifact also used to generate the TypeScript view.
            path = files("catalogmx.data").joinpath("cfdi-estado.json")
            with path.open(encoding="utf-8") as handle:
                document = json.load(handle)
            cls._data = [{"code": str(item["code"])} for item in document["data"]]
            cls._by_code = {item["code"]: item for item in cls._data}

    @classmethod
    def get_estado(cls, code: str) -> dict | None:
        cls._load_data()
        return cls._by_code.get(code.upper())

    @classmethod
    def is_valid(cls, code: str) -> bool:
        return cls.get_estado(code) is not None

    @classmethod
    def get_all(cls) -> list[dict]:
        cls._load_data()
        return cls._data.copy()

File: packages/typescript/src/catalogs/sat/cfdi_4/estado.ts
import { loadCatalogObject } from '../../../utils/catalog-loader';

export interface EstadoCfdi {
  code: string;
}

/** SAT c_Estado identifiers used by CFDI and Nómina 1.2. */
export class EstadoCfdiCatalog {
  private static _data: EstadoCfdi[] | null = null;

  private static getData(): EstadoCfdi[] {
    if (this._data === null) {
      this._data = loadCatalogObject<EstadoCfdi>('sat/cfdi_4.0/estado.json');
    }
    return this._data;
  }

  static getEstado(code: string): EstadoCfdi | undefined {
    return this.getData().find((estado) => estado.code === code.toUpperCase());
  }

  static isValid(code: string): boolean {
    return this.getEstado(code) !== undefined;
  }

  static getAll(): EstadoCfdi[] {
    return [...this.getData()];
  }
}
